Content.Server/Objectives/ObjectivesManager.cs

```diff
@@ -0,0 +1,53 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using Content.Server.Mobs;
+using Content.Server.Objectives.Interfaces;
+using Robust.Shared.Interfaces.Random;
+using Robust.Shared.IoC;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Random;
+using Robust.Shared.Utility;
+
+namespace Content.Server.Objectives
+{
+    public class ObjectivesManager : IObjectivesManager
+    {
+        [Dependency] private IPrototypeManager _prototypeManager = default!;
+        [Dependency] private IRobustRandom _random = default!;
+
+        public ObjectivePrototype[] GetAllPossibleObjectives(Mind mind)
+        {
+            return _prototypeManager.EnumeratePrototypes<ObjectivePrototype>().Where(objectivePrototype => objectivePrototype.CanBeAssigned(mind)).ToArray();
+        }
+
+        public ObjectivePrototype[] GetRandomObjectives(Mind mind, float maxDifficulty = 3)
+        {
+            var objectives = GetAllPossibleObjectives(mind);
+
+            //to prevent endless loops
+            if(objectives.Length == 0 || objectives.Sum(o => o.Difficulty) == 0f) return objectives;
+
+            var result = new List<ObjectivePrototype>();
+            var currentDifficulty = 0f;
+            while (currentDifficulty < maxDifficulty)
+            {
+                _random.Shuffle(objectives);
+                foreach (var objective in objectives)
+                {
+                    if (!_random.Prob(objective.Probability)) continue;
+
+                    result.Add(objective);
+                    currentDifficulty += objective.Difficulty;
+                    if (currentDifficulty >= maxDifficulty) break;
+                }
+            }
```

Comment:
IMO should just shuffle once based on probability (see how stationevents do it probably) and then iterate through, adding each one.